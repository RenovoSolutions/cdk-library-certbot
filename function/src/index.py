# Modified from original gist https://gist.github.com/arkadiyt/5d764c32baa43fc486ca16cb8488169a

import boto3
import certbot.main
import datetime
import os
import subprocess
from cryptography import x509
from cryptography.hazmat.backends import default_backend
from botocore.exceptions import ClientError

def store_in_secrets_manager(secret_name, secret_string):
  client = boto3.client('secretsmanager')
  try:
    create_args = {
      'Name': secret_name,
      'SecretString': secret_string
    }

    if 'CUSTOM_KMS_KEY_ID' in os.environ:
      create_args['KmsKeyId'] = os.environ['CUSTOM_KMS_KEY_ID']

    client.create_secret(**create_args)
  except ClientError as e:
    if e.response['Error']['Code'] == 'ResourceExistsException':
      update_args = {
        'SecretId': secret_name,
        'SecretString': secret_string
      }

      if 'CUSTOM_KMS_KEY_ID' in os.environ:
        update_args['KmsKeyId'] = os.environ['CUSTOM_KMS_KEY_ID']

      client.update_secret(**update_args)

def store_in_parameter_store(param_name, param_value):
  ssm = boto3.client('ssm')
  put_param_args = {
    'Name': param_name,
    'Value': param_value,
    'Type': 'SecureString',
    'Overwrite': True
  }

  if 'CUSTOM_KMS_KEY_ID' in os.environ:
    put_param_args['KeyId'] = os.environ['CUSTOM_KMS_KEY_ID']

  ssm.put_parameter(**put_param_args)

def upload_to_s3(local_path, keyname):
  print(f'INFO: Uploading {keyname} to S3')
  s3 = boto3.resource('s3')
  with open(local_path, 'rb') as file:
    data = file.read()
    s3.Bucket(os.environ['CERTIFICATE_BUCKET']).put_object(Key=f"{os.environ['OBJECT_PREFIX']}{keyname}", Body=data)

def read_and_delete_file(path, filename, storage_method):
  if not os.getenv("DRY_RUN", 'False').lower() in ["true", "1"]:
    with open(path, 'rb') as file:
      contents = file.read()
    
    try:
      string_contents = contents.decode('utf-8')
    except UnicodeDecodeError:
      print(f"Error: The file {filename} contains binary data that can't be decoded as UTF-8.")
      raise

    if storage_method == 's3':
      upload_to_s3(path, filename)
    elif storage_method == 'secretsmanager':
      store_in_secrets_manager(os.environ['CERTIFICATE_SECRET_PATH'] + filename, string_contents)
    elif storage_method == 'ssm_secure':
      store_in_parameter_store(os.environ['CERTIFICATE_PARAMETER_PATH'] + filename, string_contents)

    os.remove(path)
    return contents
  else:
    print(f'WARN: Dry run was used so {filename} was not generated.')

def provision_cert(email, domains, storage_method):
  cerbot_args = [
    'certonly',                             # Obtain a cert but don't install it
    '-n',                                   # Run in non-interactive mode
    '--agree-tos',                          # Agree to the terms of service,
    '--email', email,                       # Email
    '--dns-route53',                        # Use dns challenge with route53
    '-d', domains,                          # Domains to provision certs for
    # Override directory paths so script doesn't have to be run as root
    '--config-dir', '/tmp/config-dir/',
    '--work-dir', '/tmp/work-dir/',
    '--logs-dir', '/tmp/logs-dir/',
    '--preferred-chain', os.environ['PREFERRED_CHAIN'],
  ]
  if os.getenv("DRY_RUN", 'False').lower() in ["true", "1"]:
    print('WARN: Dry run was used so --dry-run was added to certbot args.')
    cerbot_args.append("--dry-run")

  certbot.main.main(cerbot_args)

  first_domain = domains.split(',')[0]
  path = '/tmp/config-dir/live/' + first_domain + '/'
  return {
    'certificate': read_and_delete_file(path + 'cert.pem', 'cert.pem', storage_method),
    'private_key': read_and_delete_file(path + 'privkey.pem', 'privkey.pem', storage_method),
    'certificate_chain': read_and_delete_file(path + 'chain.pem', 'chain.pem', storage_method)
  }

def should_provision(domains):
  existing_cert = find_existing_cert(domains)
  if existing_cert:
    print('INFO: Cert already exists. Checking date for reissue.')
    now = datetime.datetime.now(datetime.timezone.utc)
    not_after = existing_cert['Certificate']['NotAfter']
    reissue = (not_after - now).days <= int(os.environ['REISSUE_DAYS'])
    if reissue:
      print(f'INFO: Cert will expire sometime in the next {os.environ["REISSUE_DAYS"]} days so will be reissued.')
      return reissue
    else:
      print(f'INFO: Cert wont expire in next {os.environ["REISSUE_DAYS"]} days so will NOT be reissued.')
      return reissue
  else:
    print('INFO: Cert not found in ACM. Will issue new cert.')
    return True

def find_existing_cert(domains):
  domains = frozenset(domains.split(','))

  client = boto3.client('acm')
  paginator = client.get_paginator('list_certificates')
  iterator = paginator.paginate(PaginationConfig={'MaxItems':1000})

  for page in iterator:
    for cert in page['CertificateSummaryList']:
      cert = client.describe_certificate(CertificateArn=cert['CertificateArn'])
      sans = frozenset(cert['Certificate']['SubjectAlternativeNames'])
      if sans.issubset(domains):
        return cert

  return None

def get_cert_info(certificate):
  cert = x509.load_pem_x509_certificate(certificate, default_backend())
  # could technically dig in and get all key info here, but this is the basics
  cert_info = f"""
Certificate info:
  Serial Number: {cert.serial_number}
  Issuer: {cert.issuer.rfc4514_string()}
  Validity:
    Not Before: {cert.not_valid_before}
    Not After: {cert.not_valid_after}
  Subject: {cert.subject.rfc4514_string()}
  Subject Alternative Names: {" ".join([n.value for n in cert.extensions[6].value])}
"""
  return cert_info


def notify_via_sns(topic_arn, domains, certificate):
  print('INFO: Sending SNS notification')
  cert = get_cert_info(certificate)

  client = boto3.client('sns')
  client.publish(TopicArn=topic_arn,
    Subject='Issued new LetsEncrypt certificate',
    Message='Issued new certificates for domains: ' + domains + '\n\n' + cert,
  )

def upload_cert_to_acm(cert, domains):
  print('INFO: Importing cert to ACM')
  existing_cert = find_existing_cert(domains)
  certificate_arn = existing_cert['Certificate']['CertificateArn'] if existing_cert else None

  client = boto3.client('acm')
  if certificate_arn == None:
    acm_response = client.import_certificate(
      Certificate=cert['certificate'],
      PrivateKey=cert['private_key'],
      CertificateChain=cert['certificate_chain']
    )
  else:
    acm_response = client.import_certificate(
      CertificateArn=certificate_arn,
      Certificate=cert['certificate'],
      PrivateKey=cert['private_key'],
      CertificateChain=cert['certificate_chain']
    )

  return None if certificate_arn else acm_response['CertificateArn']

def handler(event, context):
  storage_method = os.getenv('CERTIFICATE_STORAGE', 's3').lower()

  print("CERTIFICATE_STORAGE: " + storage_method)
  print("LETSENCRYPT_DOMAINS: " + os.environ['LETSENCRYPT_DOMAINS'])
  print("LETSENCRYPT_EMAIL: " + os.environ['LETSENCRYPT_EMAIL'])
  print("PREFERRED_CHAIN: " + os.environ['PREFERRED_CHAIN'])
  print("DRY_RUN: " + os.environ['DRY_RUN'])

  # Check for required environment variables based on storage method
  if storage_method == 's3' and 'CERTIFICATE_BUCKET' not in os.environ:
    raise ValueError("S3 storage selected but CERTIFICATE_BUCKET is not set")
  elif storage_method == 'secretsmanager' and 'CERTIFICATE_SECRET_PATH' not in os.environ:
    raise ValueError("Secrets Manager storage selected but CERTIFICATE_SECRET_PATH is not set")
  elif storage_method == 'ssm_secure' and 'CERTIFICATE_PARAMETER_PATH' not in os.environ:
    raise ValueError("Parameter Store storage selected but CERTIFICATE_PARAMETER_PATH is not set")

  domains = os.environ['LETSENCRYPT_DOMAINS']
  if should_provision(domains):
    cert = provision_cert(os.environ['LETSENCRYPT_EMAIL'], domains, storage_method)
    if not os.getenv("DRY_RUN", 'False').lower() in ["true", "1"]:
      upload_cert_to_acm(cert, domains)
      notify_via_sns(os.environ['NOTIFICATION_SNS_ARN'], domains, cert['certificate'])
    else:
      print('WARN: Dry run was used so ACM import and storage upload arent tested.')
