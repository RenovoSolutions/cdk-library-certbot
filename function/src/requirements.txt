boto3 >= 1.40.6
certbot >= 4.2.0
certbot-dns-route53 >= 4.2.0
cryptography >= 45.0.6
