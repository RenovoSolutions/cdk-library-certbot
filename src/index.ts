import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';

export interface ICertbotProps {
  /**
   * The comma delimited list of domains for which the Let's Encrypt certificate will be valid. Primary domain should be first.
   */
  letsencryptDomains: string;
  /**
   * The email to associate with the Let's Encrypt certificate request.
   */
  letsencryptEmail: string;
  /**
   * Any additional Lambda layers to use with the created function. For example Lambda Extensions
   */
  layers?: lambda.LayerVersion[];
  /**
   * The S3 bucket to place the resulting certificates in. If no bucket is given one will be created automatically.
   */
  bucket?: s3.Bucket;
  /**
   * The prefix to apply to the final S3 key name for the certificates. Default is no prefix.
   */
  objectPrefix?: string;
  /**
   * The numbers of days left until the prior cert expires before issuing a new one. Default is 30 days
   */
  reIssueDays?: number;
  /**
   * Set the preferred certificate chain. Default None.
   */
  preferredChain?: string;
}

export class Certbot extends cdk.Construct {

  public readonly handler: lambda.Function

  constructor(scope: cdk.Construct, id: string, props: ICertbotProps) {
    super(scope, id);

    if (props.bucket === undefined) {
      props.bucket = new s3.Bucket(this, 'bucket', {
        objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        versioned: true,
        lifecycleRules: [{
          enabled: true,
          abortIncompleteMultipartUploadAfter: cdk.Duration.days(1),
        }],
        encryption: s3.BucketEncryption.S3_MANAGED,
        enforceSSL: true,
        blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      });
    }

    props.layers = (props.layers === undefined) ? [] : props.layers;

    const functionDir = path.join(__dirname, '../function');

    this.handler = new lambda.Function(this, 'handler', {
      runtime: lambda.Runtime.PYTHON_3_8,
      code: lambda.Code.fromAsset(functionDir + '/function.zip', {
        // Not working with github actions
        // bundling: {
        //   image: lambda.Runtime.PYTHON_3_8.bundlingImage,
        //   local: {
        //     tryBundle(outputDir: string) {
        //       try {
        //         execSync('pip3 --version | grep "python 3.8"');
        //       } catch {
        //         return false;
        //       }

        //       try {
        //         execSync(`pip install -r ${path.join(functionDir, 'requirements.txt')} -t ${path.join(outputDir)}`);
        //       } catch {
        //         return false;
        //       }

        //       try {
        //         execSync(`cp -au ${functionDir}/* ${path.join(outputDir)}`);
        //       } catch {
        //         return false;
        //       }

        //       return true;
        //     },
        //   },
        // },
      }),
      handler: 'index.handler',
      environment: {
        LETSENCRYPT_DOMAINS: props.letsencryptDomains,
        LETSENCRYPT_EMAIL: props.letsencryptEmail,
        CERTIFICATE_BUCKET: props.bucket.bucketArn,
        OBJECT_PREFIX: (props.objectPrefix === undefined) ? '' : props.objectPrefix,
        REISSUE_DAYS: (props.reIssueDays === undefined) ? '30' : String(props.reIssueDays),
        PREFERRED_CHAIN: (props.preferredChain === undefined) ? 'None' : props.preferredChain,
      },
      layers: props.layers,
    });
  }
}
