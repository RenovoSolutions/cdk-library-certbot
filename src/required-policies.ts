import {
  aws_iam as iam,
  aws_sns as sns,
  Stack,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

interface RequiredPoliciesProps {
  /**
   * The role to assign policies to
   */
  readonly role: iam.Role;
  /**
   * The SNS topic to send notifications to
   */
  readonly snsTopic: sns.Topic;
  /**
   * The hostedZones that will be required for DNS verification with certbot
   */
  readonly hostedZones: string[];
};

export function assignRequiredPoliciesToRole(scope: Construct, props: RequiredPoliciesProps): void {
  props.role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'));

  props.role.addManagedPolicy(new iam.ManagedPolicy(scope, 'snsPolicy', {
    description: 'Allow the Certbot function to notify an SNS topic upon completion.',
    statements: [
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['sns:Publish'],
        resources: [props.snsTopic.topicArn],
      }),
    ],
  }));

  props.role.addManagedPolicy(new iam.ManagedPolicy(scope, 'r53Policy', {
    description: 'Allow the Certbot function to perform DNS verification.',
    statements: [
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['route53:ListHostedZones'],
        resources: ['*'],
      }),
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: [
          'route53:GetChange',
          'route53:ChangeResourceRecordSets',
        ],
        resources: ['arn:aws:route53:::change/*'].concat(props.hostedZones),
      }),
    ],
  }));

  props.role.addManagedPolicy(new iam.ManagedPolicy(scope, 'acmPolicy', {
    description: 'Allow the Certbot function to import and list certificates.',
    statements: [
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: [
          'acm:ListCertificates',
          'acm:ImportCertificate',
        ],
        resources: ['*'],
      }),
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['acm:DescribeCertificate'],
        resources: ['arn:aws:acm:' + Stack.of(scope).region + ':' + Stack.of(scope).account + ':certificate/*'],
      }),
    ],
  }));
};