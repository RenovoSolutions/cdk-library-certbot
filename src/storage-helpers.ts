import {
  aws_iam as iam,
  aws_kms as kms,
  Stack,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

interface SecretsManagerStorageProps {
  /**
   * The role to attach the Secrets Manager policy to
   */
  readonly role: iam.Role;
  /**
   * The path to store the certificates in AWS Secrets Manager
   */
  readonly secretsManagerPath: string;
  /**
   * The KMS key alias to use for encryption of the certificates in Secrets Manager
   *
   * @default `alias/aws/secretsmanager`
   */
  readonly kmsKeyAlias?: string;
};

export function configureSecretsManagerStorage(scope: Construct, props: SecretsManagerStorageProps): void {
  const keyAlias = props.kmsKeyAlias || 'alias/aws/secretsmanager';
  const keyArn = kms.Alias.fromAliasName(scope, 'kmsKeyAlias', keyAlias).keyArn;
  props.role.addManagedPolicy(new iam.ManagedPolicy(scope, 'secretsManagerPolicy', {
    statements: [
      new iam.PolicyStatement({
        actions: [
          'secretsmanager:CreateSecret',
          'secretsmanager:DeleteSecret',
          'secretsmanager:DescribeSecret',
          'secretsmanager:GetSecretValue',
          'secretsmanager:ListSecrets',
          'secretsmanager:PutSecretValue',
          'secretsmanager:UpdateSecret',
        ],
        resources: [
          `arn:aws:secretsmanager:${Stack.of(scope).region}:${Stack.of(scope).account}:secret:${props.secretsManagerPath}*`,
        ],
      }),
      new iam.PolicyStatement({
        actions: [
          'kms:Decrypt',
          'kms:Encrypt',
        ],
        resources: [
          keyArn,
        ],
      }),
    ],
  }));
}

interface SsmStorageProps {
  /**
   * The role to attach the Secrets Manager policy to
   */
  readonly role: iam.Role;
  /**
   * The path to store the certificates in Parameter Store
   */
  readonly parameterStorePath: string;
  /**
   * The KMS key alias to use for encryption of the certificates in Secrets Manager
   *
   * @default `alias/aws/secretsmanager`
   */
  readonly kmsKeyAlias?: string;
};

export function configureSSMStorage(scope: Construct, props: SsmStorageProps): void {
  const keyAlias = props.kmsKeyAlias || 'alias/aws/ssm';
  const keyArn = kms.Alias.fromAliasName(scope, 'kmsKeyAlias', keyAlias).keyArn;
  props.role.addManagedPolicy(new iam.ManagedPolicy(scope, 'ssmPolicy', {
    statements: [
      new iam.PolicyStatement({
        actions: [
          'ssm:PutParameter',
        ],
        resources: [
          `arn:aws:ssm:${Stack.of(scope).region}:${Stack.of(scope).account}:parameter${props.parameterStorePath}*`,
        ],
      }),
      new iam.PolicyStatement({
        actions: [
          'kms:Decrypt',
          'kms:Encrypt',
        ],
        resources: [
          keyArn,
        ],
      }),
    ],
  }));
}