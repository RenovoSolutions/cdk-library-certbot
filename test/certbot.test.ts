import { expect as expectCDK, countResources, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { Certbot } from '../src/index';

jest.setSystemTime(new Date('2020-01-01').getTime());

test('Snapshot', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack', {
    env: {
      account: '123456789012', // not a real account
      region: 'us-east-1',
    },
  });

  new Certbot(stack, 'Certbot', {
    letsencryptDomains: 'test.local, www.test.local',
    letsencryptEmail: 'test@test.local',
    hostedZoneNames: ['example.com'],
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('Default', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack', {
    env: {
      account: '123456789012', // not a real account
      region: 'us-east-1',
    },
  });

  new Certbot(stack, 'Certbot', {
    letsencryptDomains: 'test.local',
    letsencryptEmail: 'test@test.local',
    hostedZoneNames: ['example.com'],
  });

  expectCDK(stack).to(countResources('AWS::Lambda::Function', 1));
  expectCDK(stack).to(countResources('AWS::Events::Rule', 2)); // one for ongoing checks and one for immediate creation
  expectCDK(stack).to(countResources('AWS::S3::Bucket', 1));
  expectCDK(stack).to(countResources('AWS::IAM::ManagedPolicy', 3)); // acm, sns, and r53
  expectCDK(stack).to(countResources('AWS::IAM::Policy', 1)); // 1 inline policy for granting bucket write
  expectCDK(stack).to(countResources('AWS::IAM::Role', 1));
  expectCDK(stack).to(countResources('AWS::SNS::Topic', 1));
  expectCDK(stack).to(countResources('AWS::SNS::Subscription', 1));
});
