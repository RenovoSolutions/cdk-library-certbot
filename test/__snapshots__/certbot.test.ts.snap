// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersc99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52cArtifactHash23964726": Object {
      "Description": "Artifact hash for asset \\"c99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52c\\"",
      "Type": "String",
    },
    "AssetParametersc99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52cS3Bucket0ABE1E94": Object {
      "Description": "S3 bucket for asset \\"c99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52c\\"",
      "Type": "String",
    },
    "AssetParametersc99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52cS3VersionKey1DFF3193": Object {
      "Description": "S3 key for asset version \\"c99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52c\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CertbotacmPolicy047B0647": Object {
      "Properties": Object {
        "Description": "Allow the Certbot function to import and list certificates.",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:ListCertificates",
                "acm:ImportCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "acm:DescribeCertificate",
              "Effect": "Allow",
              "Resource": "arn:aws:acm:us-east-1:123456789012:certificate/*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "Certbotbucket4B0141F9": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "AbortIncompleteMultipartUpload": Object {
                "DaysAfterInitiation": 1,
              },
              "Status": "Enabled",
            },
          ],
        },
        "OwnershipControls": Object {
          "Rules": Array [
            Object {
              "ObjectOwnership": "BucketOwnerPreferred",
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "CertbotbucketPolicy2B76343E": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "Certbotbucket4B0141F9",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "Certbotbucket4B0141F9",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "Certbotbucket4B0141F9",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CertbothandlerD1E99A6A": Object {
      "DependsOn": Array [
        "CertbotroleDefaultPolicy6B0533F6",
        "Certbotrole7718327E",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52cS3Bucket0ABE1E94",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52cS3VersionKey1DFF3193",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc99fc6a00a14f54079fc625de094b5468284f46eaf8dfdd8ff4f8cb540c9b52cS3VersionKey1DFF3193",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "Certbot Renewal Lambda for domain test.local",
        "Environment": Object {
          "Variables": Object {
            "CERTIFICATE_BUCKET": Object {
              "Ref": "Certbotbucket4B0141F9",
            },
            "LETSENCRYPT_DOMAINS": "test.local, www.test.local",
            "LETSENCRYPT_EMAIL": "test@test.local",
            "NOTIFICATION_SNS_ARN": Object {
              "Ref": "CertbottopicCA144CF6",
            },
            "OBJECT_PREFIX": "",
            "PREFERRED_CHAIN": "None",
            "REISSUE_DAYS": "30",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "Certbotrole7718327E",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 180,
      },
      "Type": "AWS::Lambda::Function",
    },
    "Certbotr53PolicyAE23E6E5": Object {
      "Properties": Object {
        "Description": "Allow the Certbot function to perform DNS verification.",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "route53:ListHostedZones",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "route53:GetChange",
                "route53:ChangeResourceRecordSets",
              ],
              "Effect": "Allow",
              "Resource": Array [
                "arn:aws:route53:::change/*",
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":route53:::hostedzone/DUMMY",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "Certbotrole7718327E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Ref": "CertbotsnsPolicyB1083589",
          },
          Object {
            "Ref": "Certbotr53PolicyAE23E6E5",
          },
          Object {
            "Ref": "CertbotacmPolicy047B0647",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CertbotroleDefaultPolicy6B0533F6": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "Certbotbucket4B0141F9",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "Certbotbucket4B0141F9",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CertbotroleDefaultPolicy6B0533F6",
        "Roles": Array [
          Object {
            "Ref": "Certbotrole7718327E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CertbotsnsPolicyB1083589": Object {
      "Properties": Object {
        "Description": "Allow the Certbot function to notify an SNS topic upon completion.",
        "Path": "/",
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "CertbottopicCA144CF6",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::ManagedPolicy",
    },
    "CertbottopicCA144CF6": Object {
      "Type": "AWS::SNS::Topic",
    },
    "Certbottopictesttestlocal58092F4B": Object {
      "Properties": Object {
        "Endpoint": "test@test.local",
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "CertbottopicCA144CF6",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "CertbottriggerAllowEventRuleTestStackCertbothandlerA125D3999C57ECC2": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "CertbothandlerD1E99A6A",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "CertbottriggerEEDF6DB7",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "CertbottriggerEEDF6DB7": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(0 0 ? * 1 *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "CertbothandlerD1E99A6A",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "CertbottriggerImmediate08D06D4E": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(10 0 3 1 ? 2020)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "CertbothandlerD1E99A6A",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "CertbottriggerImmediateAllowEventRuleTestStackCertbothandlerA125D39916D7FE46": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "CertbothandlerD1E99A6A",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "CertbottriggerImmediate08D06D4E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
